crates/core/src/execution_context.rs

```diff
@@ -1,17 +1,126 @@
+use std::{cell::RefCell};
+
 use derive_more::Display;
 use spacetimedb_lib::Address;
+use spacetimedb_primitives::TableId;
+
+use crate::db::db_metrics::DB_METRICS;
+
+pub enum MetricType {
+    RdbNumIndexSeeks,
+    RdbNumKeysScanned,
+    RdbNumRowsFetched,
+}
+#[derive(Default, Clone)]
+struct BufferMetric {
+    pub table_id: TableId,
+    pub rdb_num_index_seeks: u64,
+    pub rdb_num_keys_scanned: u64,
+    pub rdb_num_rows_fetched: u64,
+    pub cache_table_name: String,
+}
+
+impl BufferMetric {
+    pub fn inc_by(&mut self, ty: MetricType, val: u64) {
+        match ty {
+            MetricType::RdbNumIndexSeeks => {
+                self.rdb_num_index_seeks += val;
+            }
+            MetricType::RdbNumKeysScanned => {
+                self.rdb_num_keys_scanned += val;
+            }
+            MetricType::RdbNumRowsFetched => {
+                self.rdb_num_rows_fetched += val;
+            }
+        }
+    }
+}
+
+impl BufferMetric {
+    pub fn new(table_id: TableId, table_name: &str) -> Self {
+        Self {
+            table_id,
+            rdb_num_index_seeks: 0,
+            rdb_num_keys_scanned: 0,
+            rdb_num_rows_fetched: 0,
+            cache_table_name: table_name.to_string(),
+        }
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct Metrics(Vec<BufferMetric>);
+impl Metrics {
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    pub fn add_metric(&mut self, table_id: TableId, table_name: &str) {
+        self.0.push(BufferMetric::new(table_id, table_name));
+    }
+
+    pub fn inc_by(&mut self, table_id: TableId, ty: MetricType, val: u64) {
+        if let Some(metric) = self.0.iter_mut().find(|x| x.table_id == table_id) {
```

Comment:
I tested this, and it doesn't add that much overhead to the benchmark (~1.5%). But that benchmark only reads from two tables. I am a little concerned about transactions where this vec could be much larger.